### PR TITLE
FLW-579 iOS: Elected validators in Era service

### PR DESCRIPTION
**Pull request by ERussel**
- era validation service implementation add: main purpose is to fetch validators list when era changes and provide it to other services on request
- refactor service loading by introducing service coordinator and moving initialization of the services to the correponding factories
- add operations to create storage keys based on runtime: MapKeyEncodingOperation, DoubleMapKeyEncodingOperation
- add storage response decoding operation based on runtime
- add storage request factory to incapsulate pipeline: prepare key, get data, decode data

### CODE REVIEW

File: fearless/Common/Operation/StorageDecodingOperation.swift [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import FearlessUtils
+import RobinHood
+
+enum StorageDecodingOperationError: Error {
+    case missingRequiredParams
+    case invalidStoragePath
+}
+
+final class StorageDecodingOperation<T: Decodable>: BaseOperation<T> {
+    var data: Data?
```
Comment (ef1rspb) [inline fearless/Common/Operation/StorageDecodingOperation.swift:11]: let?

Comment (ERussel) [reply fearless/Common/Operation/StorageDecodingOperation.swift:11]: This dependency can come from another operation, so we need an opportunity to update this in config closure

File: fearless/Common/Operation/StorageRequestFactory.swift [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            } catch {
+                decodingOperation.result = .failure(error)
+            }
+        }
+
+        decodingOperation.addDependency(queryOperation)
+
+        let mapOperation = ClosureOperation<[StorageResponse<T>]> {
+            let result = try queryOperation.extractNoCancellableResultData()
+            let keysData = result
```
Comment (ef1rspb) [inline fearless/Common/Operation/StorageRequestFactory.swift:74]: result.flatMap({ StorageUpdateData(update: $0).changes }) is computed 3 times. Suggest define 
```suggestion
            let resultChangesData = result.flatMap { StorageUpdateData(update: $0).changes }
```

File: fearless/Common/Operation/StorageRequestFactory.swift [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                                        dependencies: dependencies)
+    }
+
+    func queryItems<K, T>(engine: JSONRPCEngine,
+                          keyParams: @escaping  () throws -> [K],
+                          factory: @escaping  () throws -> RuntimeCoderFactoryProtocol,
+                          storagePath: StorageCodingPath)
+    -> CompoundOperationWrapper<[StorageResponse<T>]> where K: Encodable, T: Decodable {
+
+        let currentRemoteFactory = remoteFactory
```
Comment (ef1rspb) [inline fearless/Common/Operation/StorageRequestFactory.swift:117]: currentRemoteFactory is used just once

File: fearless/Common/Operation/StorageRequestFactory.swift (hunk 2 above)
Comment (ERussel) [reply fearless/Common/Operation/StorageRequestFactory.swift]: Fixed

File: fearless/Common/Operation/StorageRequestFactory.swift (hunk 3 above)
Comment (ERussel) [reply fearless/Common/Operation/StorageRequestFactory.swift]: Fixed

File: fearless/Common/Services/ServiceCoordinator.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import SoraKeystore
+import SoraFoundation
+
+protocol ServiceCoordinatorProtocol: ApplicationServiceProtocol {
+    func updateOnAccountChange()
+    func updateOnNetworkChange()
+}
+
+final class ServiceCoordinator {
+    let webSocketService: WebSocketServiceProtocol
```
Comment (ArsenyZ) [inline fearless/Common/Services/ServiceCoordinator.swift:11]: yeeeaah

**Review by ArsenyZ — APPROVED** [review]

**Pull request merged**